src/features/courses/course-controller.ts

```diff
@@ -2522,6 +2634,243 @@ class CourseController {
+            return await StudentGradeInstance.create({
+                userId,
+                courseWWTopicQuestionId: courseTopicQuestionId,
+                webworkQuestionPath,
+                randomSeed,
+                problemNumber,
+            });
+        } catch (e) {
+            throw new WrappedError('Could not create new student grade instance', e);
+        }
+    }
+
+    async isQuestionAnAssessment(questionId: number): Promise<boolean> {
+        const question = await courseRepository.getQuestion({ id: questionId });
+        const topic = await courseRepository.getCourseTopic({ id: question.courseTopicContentId });
+        return topic.topicTypeId === 2;
+    }
+
+    /*
+    * Determine if a user is allowed to view a question 
+    * If the question belongs to an assessment, a studentTopicAssessmentInfoId is REQUIRED
+    */
+    async userCanViewQuestionId(user: User, questionId: number, studentTopicAssessmentInfoId?: number): Promise<boolean> {
+        if (user.roleId === Role.PROFESSOR || user.roleId === Role.ADMIN) return true;
+        const question = await courseRepository.getQuestion({ id: questionId });
+        const topic = await question.getTopic();
+        if (topic.topicTypeId === 1) {
+            return (topic.startDate.toMoment().isBefore(moment()));
+        } else if (topic.topicTypeId === 2) {
+            if (_.isNil(studentTopicAssessmentInfoId)) throw new WrappedError('Cannot tell if a user can view this question when a version id is not supplied');
+            const topicInfo = await topic.getTopicAssessmentInfo();
+            const studentTopicInfo = await this.getStudentTopicAssessmentInfoById(studentTopicAssessmentInfoId);
+            const topicIsLive = moment().isBetween(studentTopicInfo.startTime.toMoment(), studentTopicInfo.endTime.toMoment());
+            if (topicIsLive) {
+                return true;
+            } else {
+                // strike that -- reverse it ;)
+                return !topicInfo.hideProblemsAfterFinish;
+            }
+        }
+        // we should never get here - but if we do, then the answer is NO
+        return false;
+    };
+
+    scoreAssessment = (results: SubmittedAssessmentResultContext[]): ScoreAssessmentResult => {
+        let totalScore = 0;
+        let bestVersionScore = 0;
+        let bestOverallVersion = 0;
+        const problemScores: Array<number> = [];
+
+        results.forEach((result: SubmittedAssessmentResultContext) => {
+            const { questionResponse, grade, instance, weight } = result;
+            totalScore += weight * questionResponse.problem_result.score;
+            bestVersionScore += weight * instance.scoreForBestVersion;
+            bestOverallVersion += weight * grade.bestScore;
+            problemScores[result.instance.problemNumber] = result.questionResponse.problem_result.score * result.weight;
+        });
+
+        problemScores[0] = totalScore; // scores indexed by problem number, total at [0]
```

Comment:
I don't think problem scores should be an array, there is no guarantee that problem numbers have no gaps and are 1 based

This should never be the case but then again we saw issues with this assumption when we first did create course

Instead I would rather problemScores was of type `{[key: number]: number}` (dictionary of number to number)

The key for total score should be total score for readability

Reply:
{[key: string]: number}
{[key: string | number]: number}
{[key: any]: number} // Don't do this!